### PR TITLE
#163426579 -  Implement User Signup

### PR DESCRIPTION
**Pull request by ghost**
#### What does this PR do?
Create User Sign up page

#### Description of Task to be completed?
* [x] write tests for the feature
* [x] create the Input component
* [x] create the Signup page

#### How should this be manually tested?
- clone the repo
- cd into the directory
- cd into the folder and run `npm install` to install dependencies
- run `npm test`
- run `npm start`

#### What are the relevant pivotal tracker stories?
[#163426579](https://www.pivotaltracker.com/story/show/163426579)

Signup-page

### CODE REVIEW

**Review by obasajujoshua31 — APPROVED** [review]

**Review by Ndifreke — APPROVED** [review]

**Review by benfluleck — CHANGES_REQUESTED** [review]
Please implement feedback

File: __tests__/HomePage.test.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import React from "react";
 import Homepage from "@/pages/Homepage";
```
Comment (benfluleck) [inline __tests__/HomePage.test.js]: Please change this alias, this is not readable. You can create an alias for pages

**Review by benfluleck — CHANGES_REQUESTED** [review]

File: __tests__/Input.test.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    expect(wrapper.find("div").length).toEqual(2);
+  });
+
+  it("should render 1 <input> element", () => {
+    expect(wrapper.find("input").length).toEqual(1);
+  });
+
+  it("should allow us to set type props", () => {
+    const { type } = wrapper.props().children.props.children[0].props;
+    expect(wrapper.props());
```
Comment (benfluleck) [inline __tests__/Input.test.js]: this line is not testing anything?

File: __tests__/Input.test.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
+import React from "react";
+import Input from "@/components/common/Input";
```
Comment (benfluleck) [inline __tests__/Input.test.js]: I told you to redo this alias

File: __tests__/Input.test.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    expect(value).toEqual("value");
+  });
+
+  it("should allow us to set className props", () => {
+    const { className } = wrapper.props().children.props.children[0].props;
+    expect(className).toEqual("className");
+  });
+
+  it("should allow us to set handleChange props", () => {
+    const { handleChange } = wrapper.props().children.props.children[0].props;
```
Comment (benfluleck) [inline __tests__/Input.test.js]: Use jest.fn() to test that handleChange is called

File: src/assets/styles/SignUpPage.scss [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+  background-position: center;
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  flex-direction: column;
+  .sign-up-page__form {
+    margin: auto;
+    width: 100%;
+    height: 404px;
+    border: 1px solid #fff;
```
Comment (benfluleck) [inline src/assets/styles/SignUpPage.scss:17]: put the #fff in your color variables

File: src/assets/styles/SignUpPage.scss [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+      font-size: 1.2rem;
+      margin-bottom: 30px;
+    }
+    .sign-up-page__signup-link {
+      display: flex;
+      justify-content: flex-start;
+      font-size: 0.98rem;
+      a {
+        margin-left: 0.4rem;
+        color: $text;
```
Comment (benfluleck) [inline src/assets/styles/SignUpPage.scss:33]: Change this to text-colors

File: src/assets/styles/SignUpPage.scss [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+.sign-up-page__heading {
+  color: $text;
+  font-size: 1.1rem;
+  text-align: center;
+}
+
+.sign-up-page__btn {
+  background-color: $background !important;
+  color: $text !important;
```
Comment (benfluleck) [inline src/assets/styles/SignUpPage.scss]: Why are these already being set to important??

File: src/assets/styles/SignUpPage.scss [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    }
+    .sign-up-page__form-container {
+      width: 330px;
+      margin-top: 150px;
+    }
+  }
+
+  .ui .input,
+  .sign-up-page__btn {
+    height: 30px !important;
```
Comment (benfluleck) [inline src/assets/styles/SignUpPage.scss:171]: This is the first page and you are already creating important tags, this is not a good sign

File: src/index.scss [hunk 9]
```diff
@@ -2,8 +2,81 @@
...
+}
+
+.ui.large.form {
+  margin: 15px auto 0;
+  height: 100%;
+}
+
+.ui.black.basic.button {
+    
+    background: #FFFFFF;
```
Comment (benfluleck) [inline src/index.scss]: This is not indented properly

File: src/index.scss [hunk 10]
```diff
@@ -2,8 +2,81 @@
...
+}
+
+.ui.form .field {
+    margin: 0 0 0.5em;
+}
+
+.ui.fluid.left.icon.input {
+  width: 299.51px;
+  height: 37.54px;
+  background: #ffffff;
```
Comment (benfluleck) [inline src/index.scss]: All these colours on this page need to go into your colors.scss

File: src/index.scss [hunk 11]
```diff
@@ -2,8 +2,81 @@
...
   color: #9e2a2b;
 }
+
+.ui.container {
+  display: grid;
+  position: relative;
+  width: 100%;
+  height: 100%;
+  background: rgba(0, 0, 0, 0.63);
+  /* background: url(iso-republic-hands-using-laptops.jpg); */
```
Comment (benfluleck) [inline src/index.scss]: Remove this comment

File: src/index.scss [hunk 12]
```diff
@@ -2,8 +2,81 @@
 * {
   padding: 0;
   margin: 0;
+  font-family: Helvetica;
```
Comment (benfluleck) [inline src/index.scss]: create a font.scss for your fonts

**Review by benfluleck — CHANGES_REQUESTED** [review]

File: jest.config.js [hunk 13]
```diff
@@ -6,10 +6,15 @@ module.exports = {
   rootDir: projectDir,
   moduleNameMapper: {
     "@/(.*)$": path.join(projectDir, "src/$1"),
-    components: "<rootDir>/components",
+    components: "<rootDir>/src/components",
     pages: "<rootDir>/src/pages",
     src: "<rootDir>/src",
-    redux: "<rootDir>/src/redux",
+    "^<Components>(.*)$": "<rootDir>/src/components$1",
```
Comment (benfluleck) [inline jest.config.js]: `<components>` and `<assets>`

File: webpack/webpack.common.js [hunk 14]
```diff
@@ -39,13 +31,28 @@ module.exports = {
...
+          }
+        ]
       }
     ]
   },
   resolve: {
     alias: {
       "<Pages>": path.resolve(__dirname, "../src/pages"),
-      "<Components>": path.resolve(__dirname, "../src/components")
+      "<Components>": path.resolve(__dirname, "../src/components"),
```
Comment (benfluleck) [inline webpack/webpack.common.js]: Please change the casing on all of these

File: __tests__/Input.test.js (hunk 4 above)
Comment (benfluleck) [reply __tests__/Input.test.js]: To achieve this you will need to find one of the inputs with wrapper.find() and then trigger the change event. Allowing the handleChange to be called cc @teepha

File: jest.config.js [hunk 15]
```diff
@@ -6,10 +6,15 @@ module.exports = {
   rootDir: projectDir,
   moduleNameMapper: {
     "@/(.*)$": path.join(projectDir, "src/$1"),
-    components: "<rootDir>/components",
+    components: "<rootDir>/src/components",
```
Comment (benfluleck) [inline jest.config.js]: Talk to Ndifreke and ask what he is doing here so you can refactor this properly

File: src/assets/styles/SignUpPage.scss [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+    }
+    .sign-up-page__signup-link {
+      display: flex;
+      justify-content: flex-start;
+      font-size: 0.98rem;
+      a {
+        margin-left: 0.4rem;
+        color: $text;
+        &:hover {
+          color: $background;
```
Comment (benfluleck) [inline src/assets/styles/SignUpPage.scss]: background-color??? maybe give it a name

File: src/assets/styles/SignUpPage.scss [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+          transition: all 0.2s ease-in-out;
+        }
+      }
+
+      img {
+        height: 30px;
+        padding-left: 1rem;
+      }
+      span {
+        color: white;
```
Comment (benfluleck) [inline src/assets/styles/SignUpPage.scss:56]: Don't mix named colors and hex colors

File: src/assets/styles/SignUpPage.scss [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+    background-color: $text;
+    padding: 8px 25px;
+    border-radius: 5px;
+    font-size: 1.1rem;
+
+    a {
+      color: $black;
+
+      &:hover {
+        color: $background;
```
Comment (benfluleck) [inline src/assets/styles/SignUpPage.scss]: Rename this to a more descriptive color

File: src/components/common/Input.js [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+    icon,
+  } = props;
+  return (
+    <div className="field">
+      <div className="ui left icon input">
+        <input
+          type={type || "text"}
+          name={name}
+          className={className}
+          placeholder={placeholder || "Enter input"}
```
Comment (benfluleck) [inline src/components/common/Input.js]: Use defaultProps for this case

File: src/components/common/Input.js [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+    className,
+    placeholder,
+    handleChange,
+    icon,
+  } = props;
+  return (
+    <div className="field">
+      <div className="ui left icon input">
+        <input
+          type={type || "text"}
```
Comment (benfluleck) [inline src/components/common/Input.js]: Use default props for this case

File: src/components/common/Input.js [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+          name={name}
+          className={className}
+          placeholder={placeholder || "Enter input"}
+          value={value}
+          onChange={handleChange}
```
Comment (benfluleck) [inline src/components/common/Input.js]: Make this required and default to an empty function in defaultProps like () => {}

File: __tests__/actions/auth/signup.test.js [hunk 22]
```diff
@@ -0,0 +1,47 @@
+import moxios from "moxios";
+import signUpUser from "<actions>/auth/signup";
+import actionTypes from "<actions>/auth/types";
+import userData from "../../__mock__/userData";
```
Comment (benfluleck) [inline __tests__/actions/auth/signup.test.js]: Please add the mocks to the alias

File: __tests__/setup.config.js [hunk 23]
```diff
@@ -1,5 +1,15 @@
...
 import Adapter from "enzyme-adapter-react-16";
+import configureMockStore from "redux-mock-store";
 
 Enzyme.configure({ adapter: new Adapter() });
-export { Enzyme };
+
+const middleware = [thunk];
+const storeMock = configureMockStore(middleware);
+
+global.React = React;
```
Comment (benfluleck) [inline __tests__/setup.config.js]: I do not agree with this you can not just attack all of these on the global object. IT already has enough objects and functions to worry about

File: jest.config.js [hunk 24]
```diff
@@ -1,15 +1,28 @@
...
+  moduleFileExtensions: ["jsx", "js", "json", "png"],
+  setupFiles: ["<rootDir>/__tests__/setup.config.js"],
   rootDir: projectDir,
   moduleNameMapper: {
-    "@/(.*)$": path.join(projectDir, "src/$1"),
-    components: "<rootDir>/components",
-    pages: "<rootDir>/src/pages",
-    src: "<rootDir>/src",
-    redux: "<rootDir>/src/redux",
+    "\\.(svg)$": "<rootDir>/__tests__/__mock__",
```
Comment (benfluleck) [inline jest.config.js]: Do you use svgs or pngs ?

File: package.json [hunk 25]
```diff
@@ -58,9 +60,20 @@
     "webpack-merge": "^4.2.1"
   },
   "dependencies": {
+    "axios": "^0.18.0",
     "dotenv": "^6.2.0",
+    "jwt-decode": "^2.2.0",
```
Comment (benfluleck) [inline package.json]: What is this doing here?

File: src/api/request.js [hunk 26]
```diff
@@ -0,0 +1,21 @@
+import axios from "axios";
+
+export default async (url, method, data = {}) => {
+  try {
+    const response = await axios({
+      headers: {
+        "Access-Control-Allow-Credentials": true,
+        "Access-Control-Allow-Origin": "*",
```
Comment (benfluleck) [inline src/api/request.js:8]: What is this file doing? I asked @obasajujoshua31  and he could not tell me

File: src/components/Routes.jsx [hunk 27]
```diff
@@ -0,0 +1,32 @@
+import HomePage from "<pages>/Homepage";
+import SignUpWrapper from "<components>/auth/SignUp";
```
Comment (benfluleck) [inline src/components/Routes.jsx]: This file is not a jsx file as it doesn't rely on React. It is more a javascript file with a routes object. So it should be `routes.js`

File: src/components/common/Input.jsx [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+Input.defaultProps = {
+  className: "",
+  type: "text",
+  placeholder: "Enter input"
+};
+
+Input.propTypes = {
+  type: PropTypes.string,
+  name: PropTypes.string,
+  value: PropTypes.string,
```
Comment (benfluleck) [inline src/components/common/Input.jsx:48]: This should be required

File: src/constants/constants.js [hunk 29]
```diff
@@ -0,0 +1,3 @@
+export const API_URL = "http://localhost:3000/api/v1";
```
Comment (benfluleck) [inline src/constants/constants.js]: should this be pointing to your local branches or heroku ?

File: src/pages/SignUpPage.jsx [hunk 30]
```diff
@@ -0,0 +1,94 @@
...
+          </p>
+          <SocialLink />
+        </form>
+      </div>
+    </div>
+  );
+};
+
+SignUpPage.propTypes = {
+  firstName: PropTypes.string,
```
Comment (benfluleck) [inline src/pages/SignUpPage.jsx]: For Signup firstName , lastName and email should be required

**Review by benfluleck — CHANGES_REQUESTED** [review]

**Review by eokwukwe — APPROVED** [review]

File: src/components/common/Input.jsx (hunk 28 above)
Comment (ghost) [reply src/components/common/Input.jsx:48]: I thought you said I should not make it required, that  I should set defaultValues instead

File: __tests__/pages/ProfilePage.test.js [hunk 31]
```diff
@@ -0,0 +1,14 @@
+import React from "react";
+import { Enzyme } from "../setup.config";
+import ProfilePage from "<pages>/ProfilePage";
+
+const { shallow } = Enzyme;
+
+describe("ProfilePage", () => {
+  it("should contain h1 element", () => {
+    const profilePageShallow = shallow(<ProfilePage />);
+    expect(profilePageShallow.contains(<h1>welcome to Profile</h1>)).toBe(
```
Comment (benfluleck) [inline __tests__/pages/ProfilePage.test.js]: Please change this line to `expect(profilePageShallow.contains(<h1>welcome to Profile</h1>)).toBeTruthy()`

File: __tests__/utils/validation.test.js [hunk 32]
```diff
@@ -0,0 +1,116 @@
...
+  it("should return errors for a form that is not valid", () => {
+    const isNotValidResult = isNotValid(state.errors);
+    expect(isNotValidResult).toEqual(false);
+  });
+
+  it("should not return any error for an unknown field", () => {
+    const validationResult = validateCreateForm(state, {
+      target: { name: "unknown", value: "gorilla" }
+    });
+    expect(validationResult.unknown).toEqual(undefined);
```
Comment (benfluleck) [inline __tests__/utils/validation.test.js]: use `.toBeFalsy`

File: src/components/auth/AuthWrapper.jsx [hunk 33]
```diff
@@ -0,0 +1,51 @@
...
+        password: "",
+        firstName: "",
+        lastName: ""
+      });
+    };
+
+    render() {
+      return (
+        <AuthComponent
+          state={this.state}
```
Comment (benfluleck) [inline src/components/auth/AuthWrapper.jsx]: Please do not treat state like props. Rename this  or destructure the arguments

File: src/components/auth/SignUp.jsx [hunk 34]
```diff
@@ -0,0 +1,57 @@
...
+
+SignUpWrapper.propTypes = {
+  firstName: PropTypes.string,
+  lastName: PropTypes.string,
+  email: PropTypes.string,
+  password: PropTypes.string,
+  handleChange: PropTypes.func,
+  handleSubmit: PropTypes.func,
+  errors: PropTypes.object,
+  state: PropTypes.object,
```
Comment (benfluleck) [inline src/components/auth/SignUp.jsx]: Remove this from the props check

File: src/components/auth/VerifyAccount.jsx [hunk 35]
```diff
@@ -0,0 +1,36 @@
...
+      .then(() => {
+        this.props.history.push("/profile");
+      })
+      .catch(() => {
+        this.props.history.push("/");
+      });
+  }
+
+  render() {
+    return null;
```
Comment (benfluleck) [inline src/components/auth/VerifyAccount.jsx]: What is this?

Comment (benfluleck) [reply src/components/auth/VerifyAccount.jsx]: This goes against React principles. This probably is better in the route as `renderProps`

**Review by benfluleck — CHANGES_REQUESTED** [review]

File: __tests__/index.js [hunk 36]
```diff
@@ -1 +1,11 @@
-import "./HomePage.test";
+import "./pages/HomePage.test";
+import "./pages/ProfilePage.test";
+// import "./components/auth/AuthWrapper.test";
```
Comment (benfluleck) [inline __tests__/index.js]: What is this doing here ?

File: __tests__/pages/ProfilePage.test.js [hunk 37]
```diff
@@ -0,0 +1,14 @@
+import React from "react";
+import { Enzyme } from "../setup.config";
+import ProfilePage from "<pages>/ProfilePage";
+
+const { shallow } = Enzyme;
+
+describe("ProfilePage", () => {
+  it("should contain h1 element", () => {
+    const profilePageShallow = shallow(<ProfilePage />);
+    expect(
```
Comment (benfluleck) [inline __tests__/pages/ProfilePage.test.js:11]: Space between the declarations and the test please

File: __tests__/reducers/authReducer.test.js [hunk 38]
```diff
@@ -0,0 +1,40 @@
...
+        }
+      )
+    ).toEqual({
+      isAuthenticated: true
+    });
+  });
+
+  it("should handle VERIFY_USER", () => {
+    const payload = "Verification Successful!";
+    expect(
```
Comment (benfluleck) [inline __tests__/reducers/authReducer.test.js:31]: Space here please

File: __tests__/reducers/authReducer.test.js [hunk 39]
```diff
@@ -0,0 +1,40 @@
...
+  it("should return the initial state ", () => {
+    expect(reducers(undefined, {})).toEqual({
+      isAuthenticated: false,
+      error: "",
+      authMessage: ""
+    });
+  });
+
+  it("should handle SET_CURRENT_USER", () => {
+    expect(
```
Comment (benfluleck) [inline __tests__/reducers/authReducer.test.js:16]: Space here please

File: __tests__/reducers/authReducer.test.js [hunk 40]
```diff
@@ -0,0 +1,40 @@
+import reducers from "../../src/reducers/authReducer";
+import actionTypes from "../../src/actions/auth/types";
+
+describe("AuthReducer test", () => {
+  it("should return the initial state ", () => {
+    expect(reducers(undefined, {})).toEqual({
```
Comment (benfluleck) [inline __tests__/reducers/authReducer.test.js:7]: Space here please

File: __tests__/utils/validation.test.js [hunk 41]
```diff
@@ -0,0 +1,116 @@
...
+  password: ""
+};
+
+describe("Create Validation Form test", () => {
+  it("should return form errors for a form that is not valid", () => {
+    const formErrors = validateCreateForm(
+      state,
+      mockChangeValue("firstName", "")
+    );
+    expect(formErrors.firstName).toEqual("First name is too short");
```
Comment (benfluleck) [inline __tests__/utils/validation.test.js:30]: Space here please

File: src/actions/auth/socialAuth.js [hunk 42]
```diff
@@ -0,0 +1,20 @@
+import request from "<api>/request";
+import actionTypes from "./types";
+
+export default (url, socialToken) => dispatch => {
+  return request(`${url}/${socialToken}`, "GET")
+    .then(response => {
+      localStorage.setItem("token", response.token);
+      dispatch({
```
Comment (benfluleck) [inline src/actions/auth/socialAuth.js:9]: Please put a space here

File: src/actions/auth/verifyUser.js [hunk 43]
```diff
@@ -0,0 +1,21 @@
+import request from "<api>/request";
+import actionTypes from "./types";
+import { API_URL } from "<constants>/constants";
+
+export default token => dispatch => {
+  return request(`${API_URL}/auth/verification/${token}`, "GET")
+    .then(response => {
+      localStorage.setItem("token", response.token);
+      dispatch({
```
Comment (benfluleck) [inline src/actions/auth/verifyUser.js:10]: Please put a space here please

File: src/components/auth/SignUp.jsx [hunk 44]
```diff
@@ -0,0 +1,57 @@
...
+
+SignUpWrapper.propTypes = {
+  firstName: PropTypes.string,
+  lastName: PropTypes.string,
+  email: PropTypes.string,
+  password: PropTypes.string,
+  handleChange: PropTypes.func,
+  handleSubmit: PropTypes.func,
+  errors: PropTypes.object,
+  authData: PropTypes.object,
```
Comment (benfluleck) [inline src/components/auth/SignUp.jsx:45]: This shape is not correct but I will just leave you

**Review by benfluleck — CHANGES_REQUESTED** [review]

**Comment by benfluleck** [comment]
Also you have a conflict

File: __tests__/components/BaseRoute.test.js [hunk 45]
```diff
@@ -0,0 +1,48 @@
...
+import { MemoryRouter } from "react-router-dom";
+import { mount } from "enzyme";
+import { ToastContainer } from "react-toastify";
+import BaseRoute from "<components>/BaseRoute";
+import SignUpWrapper from "<auth>/SignUp";
+import HomePage from "<pages>/Homepage";
+import ProfilePage from "<pages>/ProfilePage";
+import SocialRedirect from "<auth>/SocialAuth";
+import VerifyAccount from "<auth>/VerifyAccount";
+import store from "<src>/store";
```
Comment (benfluleck) [inline __tests__/components/BaseRoute.test.js]: Are you testing our actual store? You should use a mock store.

**Review by benfluleck — APPROVED** [review]

**Pull request merged**